controllers/service/service_controller.go

```diff
@@ -99,16 +108,102 @@ func (r *serviceReconciler) reconcile(ctx context.Context, req ctrl.Request) err
 	if err := r.k8sClient.Get(ctx, req.NamespacedName, svc); err != nil {
 		return client.IgnoreNotFound(err)
 	}
+	if err := r.allocatedService(ctx, svc); err != nil {
+		return err
+	}
 	stack, lb, backendSGRequired, err := r.buildModel(ctx, svc)
 	if err != nil {
 		return err
 	}
-	if lb == nil {
-		return r.cleanupLoadBalancerResources(ctx, svc, stack)
+	if lb == nil || !svc.DeletionTimestamp.IsZero() {
+		return r.cleanupLoadBalancerResources(ctx, svc, stack, lb == nil)
 	}
 	return r.reconcileLoadBalancerResources(ctx, svc, stack, lb, backendSGRequired)
 }
 
+// allocatedService allocates a stack to a service so that it can share load balancer resources with other services.
+func (r *serviceReconciler) allocatedService(ctx context.Context, svc *corev1.Service) error {
+	if !r.initialized {
+		r.lock.Lock()
+		var serviceList corev1.ServiceList
+		if err := r.k8sClient.List(ctx, &serviceList); err != nil {
+			return err
+		}
+		for _, svc := range serviceList.Items {
+			if svc.Annotations[service.LoadBalancerStackKey] != "" {
+				r.allocatedServices[svc.Annotations[service.LoadBalancerStackKey]] = map[int]bool{}
+				for _, port := range svc.Spec.Ports {
+					r.allocatedServices[svc.Annotations[service.LoadBalancerStackKey]][int(port.NodePort)] = true
+				}
+			}
+		}
+		r.initialized = true
+		r.lock.Unlock()
+	}
+	if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return nil
+	}
+	if svc.Annotations[service.LoadBalancerAllocatingPortKey] != "true" {
+		return nil
+	}
+
+	if svc.Annotations[service.LoadBalancerStackKey] != "" {
+		return nil
+	}
+
+	if !svc.DeletionTimestamp.IsZero() {
+		if _, ok := r.allocatedServices[svc.Annotations[service.LoadBalancerStackKey]]; !ok {
```

Comment:
Based on the if-statement on line 150, `svc.Annotations[service.LoadBalancerStackKey]` will always be `""`. Is that intended?

You would also need a (read) lock around reads from `r.allocatedServices` to ensure you are not reading and writing to the map at the same time.

Reply:
yeah, my bad. It will actually short-circuit the code. Move the finalize logic to the top now.

Also, right now allocateService will be protected by a global lock now since multiple service could access the code concurrently